File: 3569.maximize-count-of-distinct-primes-after-split.java
import java.util.*;

class Solution {
    static final int MAX = 100005;
    static BitSet sieve = new BitSet(MAX);
    static {
        sieve.set(2, MAX, true);
        for (int i = 2; i * i < MAX; i++) {
            if (sieve.get(i)) {
                for (int j = i * i; j < MAX; j += i) sieve.set(j, false);
            }
        }
    }

    class SegTree {
        int n;
        BitSet[] tree;
        int size;
        SegTree(int[] nums) {
            n = nums.length;
            size = 1; while (size < n) size <<= 1;
            tree = new BitSet[size << 1];
            for (int i = 0; i < (size << 1); i++) tree[i] = new BitSet(MAX);
            build(nums);
        }
        void build(int[] nums) {
            for (int i = 0; i < n; i++) {
                if (sieve.get(nums[i])) tree[i + size].set(nums[i]);
            }
            for (int i = size - 1; i > 0; i--) {
                tree[i].or(tree[i << 1]);
                tree[i].or(tree[(i << 1) | 1]);
            }
        }
        void update(int idx, int val) {
            int pos = idx + size;
            tree[pos].clear();
            if (sieve.get(val)) tree[pos].set(val);
            for (pos >>= 1; pos > 0; pos >>= 1) {
                tree[pos].clear();
                tree[pos].or(tree[pos << 1]);
                tree[pos].or(tree[(pos << 1) | 1]);
            }
        }
        BitSet query(int l, int r) {
            BitSet res = new BitSet(MAX);
            l += size; r += size;
            while (l < r) {
                if ((l & 1) == 1) { res.or(tree[l++]); }
                if ((r & 1) == 1) { res.or(tree[--r]); }
                l >>= 1; r >>= 1;
            }
            return res;
        }
    }

    public int[] maximumCount(int[] nums, int[][] queries) {
        int n = nums.length;
        SegTree seg = new SegTree(nums);
        int[] res = new int[queries.length];
        for (int qi = 0; qi < queries.length; qi++) {
            int idx = queries[qi][0], val = queries[qi][1];
            seg.update(idx, val);
            int max = 0;
            for (int k = 1; k < n; k++) {
                BitSet prefix = seg.query(0, k);
                BitSet suffix = seg.query(k, n);
                int cnt = prefix.cardinality() + suffix.cardinality();
                max = Math.max(max, cnt);
            }
            res[qi] = max;
        }
        return res;
    }
}

File: 3569.maximize-count-of-distinct-primes-after-split.py
#
# @lc app=leetcode id=3569 lang=python3
#
# [3569] Maximize Count of Distinct Primes After Split
#

# @lc code=start
from typing import List
from collections import defaultdict
import heapq

class Solution:
    def maximumCount(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        def get_min_pos(prime_pos, min_heaps, p):
            minh = min_heaps[p]
            while minh and minh[0] not in prime_pos[p]:
                heapq.heappop(minh)
            return minh[0] if minh else -1

        def get_max_pos(prime_pos, max_heaps, p):
            maxh = max_heaps[p]
            while maxh and -maxh[0] not in prime_pos[p]:
                heapq.heappop(maxh)
            return -maxh[0] if maxh else -1

        class SegTree:
            def __init__(self, n: int):
                self.n = n
                self.tree = [0] * (4 * (n + 1))
                self.lazy = [0] * (4 * (n + 1))

            def push(self, v: int, tl: int, tr: int):
                if self.lazy[v] != 0:
                    self.tree[v] += self.lazy[v]
                    if tl != tr:
                        self.lazy[2 * v] += self.lazy[v]
                        self.lazy[2 * v + 1] += self.lazy[v]
                    self.lazy[v] = 0

            def update_range(self, v: int, tl: int, tr: int, l: int, r: int, add: int):
                self.push(v, tl, tr)
                if l > r:
                    return
                if tl == l and tr == r:
                    self.lazy[v] += add
                    self.push(v, tl, tr)
                    return
                tm = (tl + tr) // 2
                self.update_range(2 * v, tl, tm, l, min(r, tm), add)
                self.update_range(2 * v + 1, tm + 1, tr, max(l, tm + 1), r, add)
                self.tree[v] = max(self.tree[2 * v], self.tree[2 * v + 1])

            def query_max(self, v: int, tl: int, tr: int, l: int, r: int) -> int:
                self.push(v, tl, tr)
                if l > r:
                    return float('-inf')
                if tl == l and tr == r:
                    return self.tree[v]
                tm = (tl + tr) // 2
                return max(
                    self.query_max(2 * v, tl, tm, l, min(r, tm)),
                    self.query_max(2 * v + 1, tm + 1, tr, max(l, tm + 1), r)
                )

        n = len(nums)
        if n == 0:
            return []
        MAXV = 100010
        is_prime = [True] * MAXV
        is_prime[0] = is_prime[1] = False
        for i in range(2, int(MAXV ** 0.5) + 1):
            if is_prime[i]:
                for j in range(i * i, MAXV, i):
                    is_prime[j] = False

        prime_pos = defaultdict(set)
        min_heaps = defaultdict(list)
        max_heaps = defaultdict(list)
        total_distinct = 0

        for i in range(n):
            v = nums[i]
            if v < MAXV and is_prime[v]:
                p = v
                old_len = len(prime_pos[p])
                prime_pos[p].add(i)
                heapq.heappush(min_heaps[p], i)
                heapq.heappush(max_heaps[p], -i)
                if old_len == 0:
                    total_distinct += 1

        seg = SegTree(n - 1)
        for p in list(prime_pos.keys()):
            if len(prime_pos[p]) >= 2:
                mn = get_min_pos(prime_pos, min_heaps, p)
                mx = get_max_pos(prime_pos, max_heaps, p)
                l = mn + 1
                r = mx
                if l <= r:
                    seg.update_range(1, 1, n - 1, l, r, 1)

        ans = []
        for idx, val in queries:
            oldv = nums[idx]
            nums[idx] = val

            # Handle removal of oldv
            if oldv < MAXV and is_prime[oldv]:
                p = oldv
                old_len = len(prime_pos[p])
                if old_len >= 2:
                    mn = get_min_pos(prime_pos, min_heaps, p)
                    mx = get_max_pos(prime_pos, max_heaps, p)
                    l = mn + 1
                    r = mx
                    if l <= r:
                        seg.update_range(1, 1, n - 1, l, r, -1)
                prime_pos[p].remove(idx)
                new_len = old_len - 1
                if new_len == 0:
                    total_distinct -= 1
                if new_len >= 2:
                    mn = get_min_pos(prime_pos, min_heaps, p)
                    mx = get_max_pos(prime_pos, max_heaps, p)
                    l = mn + 1
                    r = mx
                    if l <= r:
                        seg.update_range(1, 1, n - 1, l, r, 1)

            # Handle addition of val
            if val < MAXV and is_prime[val]:
                p = val
                old_len = len(prime_pos[p])
                if old_len >= 2:
                    mn = get_min_pos(prime_pos, min_heaps, p)
                    mx = get_max_pos(prime_pos, max_heaps, p)
                    l = mn + 1
                    r = mx
                    if l <= r:
                        seg.update_range(1, 1, n - 1, l, r, -1)
                prime_pos[p].add(idx)
                heapq.heappush(min_heaps[p], idx)
                heapq.heappush(max_heaps[p], -idx)
                new_len = old_len + 1
                if old_len == 0:
                    total_distinct += 1
                if new_len >= 2:
                    mn = get_min_pos(prime_pos, min_heaps, p)
                    mx = get_max_pos(prime_pos, max_heaps, p)
                    l = mn + 1
                    r = mx
                    if l <= r:
                        seg.update_range(1, 1, n - 1, l, r, 1)

            max_inter = seg.query_max(1, 1, n - 1, 1, n - 1)
            ans.append(total_distinct + max_inter)

        return ans

# @lc code=end